classes/user/DAO.inc.php

```diff
@@ -302,4 +303,22 @@ public function changeSitePrimaryLocale($oldLocale, $newLocale)
             }
         }
     }
+
+    /**
+     * Delete unvalidated expired users
+     * 
+     * @param object<Carbon\Carbon> $dateTillValid      The dateTime till before which user will consider expired
+     * @param array                 $excludableUsersId  The users id to exclude form delete operation
+     * 
+     * @return int The number rows affected by DB operation
+     */
+    public function deleteUnvalidatedExpiredUsers(Carbon $dateTillValid, array $excludableUsersId = []) {
+
+        return DB::table('users')
+            ->whereNull('date_validated')
+            ->whereNull('date_last_login')
+            ->where('date_registered', '<', $dateTillValid)
+            ->when( !empty($excludableUsersId), fn($query) => $query->whereNotIn('id', $excludableUsersId) )
+            ->delete();
+    }
```

Comment:
Users may have additional settings in `user_settings`, as well as `user_user_groups` and some other places. This will be limited for unvalidated users, since they can't create other data in the system until they validate. But because our system is open to being extended by plugins, it's often hard to predict what other data may exist.

The safest thing to do is to use the repository's delete method, which is responsible for removing all traces of a user, firing hooks, etc:

```php
$userIds = DB::table('users')->where(...)->...;
foreach ($userIds as $userId) {
    Repo::user()->delete(Repo::user()->get($userId));
}
```

@asmecher we should probably figure out a consistent pattern for where "related data" should get cleaned up: in the DAO or Repository. The DAO makes sense, but hooks are only fired on a Repository level, so when going straight to the DAO, a plugin which has added data can't clean it up.

Reply:
@AndrewGearhart yeah, because these users are unvalidated they won't have any submissions, discussions, stage assignments, log entries or other things. That's what a "merge" takes care of.

However, the comment here is more about data in the user entity that's tracked in more than one database table. All users, even unvalidated ones, will have entries in the `user`, `user_settings`, and `user_user_groups` tables.

Reply:
@NateWr I see the reason to use the `delete` method from the `Repo` class . Yes, even with users deleted from the `users` table , we are leaving garbage data in some other tables associated with those deleted users. I have update the implementation as per your request .

Reply:
@touhidurabir, this appears to still be calling `DAO::delete` rather than `Repo::delete`.